Develop database schema, establish connection to database and perform read/writes.

MellowYarker opened the pull request:
By developing the schema, we will have a better idea of what data is important. Currently, the database tables and their columns store information that the backend cannot provide, so in future PRs we will extend the current data structures to enable these features. Some examples include order status, time placed/updated.

In this PR, I plan to have the backend read all the records from the database on start up, and then update/write to the database when orders are submitted, filled, and cancelled.

MellowYarker committed:
Brought in postgres crate, moved db dir

MellowYarker committed:
Got database connection, will create specialized DB for this program soon.

MellowYarker committed:
renamed db folder

MellowYarker committed:
should've been part of previous commit

MellowYarker committed:
Added database operations file, now able to fill pending orders.

MellowYarker committed:
Read in market and exchange stats. Add exchange_stats table to schema

MellowYarker committed:
Can now properly read market stats

MellowYarker committed:
Added direct func for filled orders

MellowYarker committed:
Prep sim for database upgrades by commenting it out.

MellowYarker committed:
Read in users on start up. User functions like get, get_mut, and authenticate now read from DB. Also we have fetch pending and executed from DB.

MellowYarker committed:
Changed property names of FilledOrder and resolved other compiler warnings.

MellowYarker committed:
Renamed FilledOrder to Trade, which is clearer IMO.

MellowYarker committed:
Changed comment for Trade struct

MellowYarker committed:
Renamed security to symbol, should have done this a long time ago.

MellowYarker committed:
Read pending orders in at initialization time. This is not ideal, but we have to do it for now.

cacheburner commented on src/account.rs:
```diff
@@ -171,30 +347,68 @@ impl Users {
+            }
+        }
+
+        // TODO: we call get twice if it was a cache hit.
```
Lets make sure we actually resolve this problem in a Rust way. It comes up several times and there's no point in having a cache if we use it like a stupid idiot.

MellowYarker committed:
Moved db init reads from main to account + database.rs. Also:

Updated buy/sell -> BUY/SELL.
Insert new order to Orders table.
Update database's market stats after in-mem update.

MellowYarker committed:
Updated Schema. See database changes below:

We now:
        - insert newly placed order to Orders + pendingOrders (if
          necessary) table
        - update all Orders that were part of a trade
        - remove from pending orders table if order is complete
        - insert the trades that occured into ExecutedTrades table.
        - update market stats.

MellowYarker committed:
Changed exchange.trades to exchange.has_trades:

This means we no longer store past trades in memory (long term).
Rather, we read them from the database when needed, and do a quick
cache check to see if the requested market has any trades.

Also, we initialize this data structure at start up, and update
the cache whenever new trades occur.

MellowYarker committed:
PREVIOUS COMMIT HAD WRONG FILE. Also this has a bug fix.

MellowYarker committed:
Removed inserts which weren't meant to be sent.

MellowYarker committed:
Remove executed_orders from userAccount. Waste of space, we will cache the update time later or something.

MellowYarker committed:
Increased status size

MellowYarker committed:
Cancelled orders are now updated in the database.

Added further comments for future upgrades.

MellowYarker committed:
Small mutability changes

MellowYarker committed:
get & get_mut funcs no longer do DB lookups.

Since these functions require a user to be authenticated, and
authentication updates the cache, there's no need to do DB lookups.

_get_mut is different. This function does a DB lookup, but doesn't
update the cache. Instead, we just construct the user account and return
it so the calling function can generate SQL to update the DB.

MellowYarker committed:
Another db column upgrade

MellowYarker committed:
Another table upgrade.

MellowYarker committed:
apparently some companies have long names...

MellowYarker committed:
Reverting to old code, found bug in new version.

MellowYarker committed:
Added admin command to upgrade DB. also:

We now call update_state in submit_order_to_market even when
the market had no pending orders. Need to look at this soon.

MellowYarker committed:
Fixed Queries, use transactions + func for sim.

MellowYarker committed:
Keep accidentally commiting this stupid file

MellowYarker committed:
File I meant to push 2 commits ago.

MellowYarker committed:
Use the same db connection everywhere

MellowYarker commented:
At this point, I think that the goal of this branch has been completed. The entire database has been integrated with the program, we're just missing chronistic information like execution time, update time, etc.

The unfortunate result of persistence is our performance took a turn for the worst. The time the program spends on the CPU hasn't increased too drastically, but we're communicating with disk (the database) way too often. This is going to have to be addressed in another branch, caching will save the day.

I may tidy up some *todo's* here and there before merging, although some of them require large changes (ex. dropping input references to username, since moving strings is more expensive than moving u32's).

MellowYarker committed:
Moved query construction to DB API. fn names impr+

MellowYarker committed:
Cleaned comments, terminated query strings.

MellowYarker committed:
Moved fetch funcs to database API.

MellowYarker committed:
Moved more SQL to database.rs

MellowYarker committed:
Moved _get_mut database stuff to DB API.

MellowYarker committed:
Moved more sql to database from account.

MellowYarker committed:
Rather than create new markets, read from DB only!

MellowYarker committed:
Completely removed init_stats.

MellowYarker committed:
Updated return type of submit_order...

MellowYarker committed:
Don't get users on startup, just user count.

MellowYarker committed:
Remove code to load all users on startup.

MellowYarker reviewed (commented):
As mentioned later in the PR, this will be fixed once we move to IDs rather than using a hybrid approach of IDs and usernames. We have to store &mut self rather than &mut `property_of_self, which is what is causing the issue. Essentially, this will be resolved when the parser gets better, and this will happen when the frontend is less of a hassle to deal with (saves user details and sends ID instead of username).
